evm/src/main/java/org/hyperledger/besu/evm/operation/SelfDestructOperation.java

```diff
@@ -27,54 +27,77 @@
+    this.eip6780Semantics = eip6780Semantics;
   }
 
   @Override
   public OperationResult execute(final MessageFrame frame, final EVM evm) {
-    final Address recipientAddress = Words.toAddress(frame.popStackItem());
-
+    // fist calculate cost.  There's a bit of yak shaving getting values to calculate the cost
+    final Address beneficiaryAddress = Words.toAddress(frame.popStackItem());
     // because of weird EIP150/158 reasons we care about a null account, so we can't merge this.
-    final Account recipientNullable = frame.getWorldUpdater().get(recipientAddress);
-    final Wei inheritance = frame.getWorldUpdater().get(frame.getRecipientAddress()).getBalance();
+    final Account beneficiaryNullable = frame.getWorldUpdater().get(beneficiaryAddress);
+    final boolean beneficiaryIsWarm =
+        frame.warmUpAddress(beneficiaryAddress) || gasCalculator().isPrecompile(beneficiaryAddress);
 
-    final boolean accountIsWarm =
-        frame.warmUpAddress(recipientAddress) || gasCalculator().isPrecompile(recipientAddress);
+    final Address originatorAddress = frame.getRecipientAddress();
+    final Wei originatorBalance = frame.getWorldUpdater().get(originatorAddress).getBalance();
 
     final long cost =
-        gasCalculator().selfDestructOperationGasCost(recipientNullable, inheritance)
-            + (accountIsWarm ? 0L : gasCalculator().getColdAccountAccessCost());
+        gasCalculator().selfDestructOperationGasCost(beneficiaryNullable, originatorBalance)
+            + (beneficiaryIsWarm ? 0L : gasCalculator().getColdAccountAccessCost());
 
+    // with the cost we can test for two early exits: static or not enough gas
     if (frame.isStatic()) {
       return new OperationResult(cost, ExceptionalHaltReason.ILLEGAL_STATE_CHANGE);
     } else if (frame.getRemainingGas() < cost) {
       return new OperationResult(cost, ExceptionalHaltReason.INSUFFICIENT_GAS);
     }
 
-    final Address address = frame.getRecipientAddress();
-    final MutableAccount account = frame.getWorldUpdater().getAccount(address).getMutable();
-
-    frame.addSelfDestruct(address);
+    // We passed preliminary checks, get mutable accounts.
+    final MutableAccount originatorAccount =
+        frame.getWorldUpdater().getAccount(originatorAddress).getMutable();
+    final MutableAccount beneficiaryAccount =
+        frame.getWorldUpdater().getOrCreate(beneficiaryAddress).getMutable();
 
-    final MutableAccount recipient =
-        frame.getWorldUpdater().getOrCreate(recipientAddress).getMutable();
+    // Do the "sweep," all modes send all originator balance to the beneficiary account.
+    originatorAccount.decrementBalance(originatorBalance);
+    beneficiaryAccount.incrementBalance(originatorBalance);
 
-    if (!account.getAddress().equals(recipient.getAddress())) {
-      recipient.incrementBalance(account.getBalance());
+    // If we are actually destroying the contract (pre-Cancun or same-tx-create) we need to
```

Comment:
for consistency with the rest of the comment/variable names

```suggestion
    // If we are actually destroying the originator account (pre-Cancun or same-tx-create) we need to
```